src/tools/tools/process.fs

```diff
@@ -1171,10 +1171,130 @@ type internal ThreadPoolWait internal (id:int) as this =
+            let handleCollection = ThreadPoolWait.WaitHandleCollection.GetOrAdd( handle, fun _ -> ConcurrentDictionary())
+            handleCollection.Item(continuationObject) <- ( rwh, infoFunc, ref (continuation, unblockHandle), ref DateTime.MinValue.Ticks )
+    static member private CallBack( state: Object ) (timeout:bool) = 
+        if not timeout then
+            try
+                let handle, continuationObject = state :?> (WaitHandle*Object)
+                let handleCollectionRef = ref Unchecked.defaultof<_>
+                let tuple = ref Unchecked.defaultof<_>
+                let failedReason = ref "None" 
+                let bFind = SpinWait.SpinUntil( ( fun _ ->  let bExist, handleCollection = ThreadPoolWait.WaitHandleCollection.TryGetValue( handle )
+                                                            handleCollectionRef := handleCollection
+                                                            if bExist then 
+                                                                let bExist, ret = handleCollection.TryGetValue( continuationObject )
+                                                                if bExist then 
+                                                                    tuple := ret
+                                                                    true
+                                                                else
+                                                                    failedReason := "Unable to find continuation Object"
+                                                                    false
+                                                            else
+                                                                failedReason := "Unable to find handle"
+                                                                false ),
+                                                    ThreadPoolWait.MaxSpinMs )
+                if bFind then 
+                    let rwh, infoFunc, workRef, refStartContinuation = !tuple
+                    refStartContinuation := DateTime.UtcNow.Ticks 
+                    let continuation, unblockHandle = !workRef
+                    // Long executing continuation may be flagged for performance issue. 
+                    continuation() 
+                    if Utils.IsNotNull unblockHandle  then 
+                        // If there is an unblock handle, set it. 
+                        unblockHandle.Set() |> ignore  
+                    (!handleCollectionRef).TryRemove( continuationObject ) |> ignore 
+                    if (!handleCollectionRef).IsEmpty then 
+                        ThreadPoolWait.WaitHandleCollection.TryRemove( handle) |> ignore
+                    rwh.Unregister(null) |> ignore 
+                else
+                    failwith (sprintf "ThreadPoolWait.Callback, can't find the entry of waithandle and continuation object after %d ms, reason: %s " ThreadPoolWait.MaxSpinMs !failedReason )
```

Comment:
This is unrecoverable error. On slow system, it can totally happen. However, there's nothing wrong about it, I think we should not fail here. 

I think if we handle "rwh" as the example in the email, then we can add the item to the handleCollection before calling RegisterWaitForSingleObject.  As a result, there is no need to spin here, it will be a deterministic behavior. 

BTW: I assume the purpose WaitHandleCollection is to help on monitoring?